File: c/rna-transcription/rna_transcription.c
#include <malloc.h>
#include <string.h>
#include "rna_transcription.h"

char *to_rna(const char *dna)
{
	size_t len = strlen(dna);

	char *rna = malloc(len + 1);
	if (!rna)
		return NULL;

	for (size_t i = 0; i < len; i++)
	{
		if (dna[i] == 'G')
			rna[i] = 'C';
		else if (dna[i] == 'C')
			rna[i] = 'G';
		else if (dna[i] == 'T')
			rna[i] = 'A';
		else if (dna[i] == 'A')
			rna[i] = 'U';
	}

	rna[len] = '\0';

	return rna;
}

File: zig/rna-transcription/rna_transcription.zig
const std = @import("std");
const Allocator = std.mem.Allocator;

pub fn toRna(allocator: Allocator, dna: []const u8) Allocator.Error![]const u8 {
    const rna = try allocator.dupe(u8, dna);

    for (rna) |*c| {
        c.* = switch (c.*) {
            'G' => 'C',
            'C' => 'G',
            'T' => 'A',
            'A' => 'U',
            else => continue,
        };
    }

    return rna;
}
